container/src/lib.rs

```diff
@@ -140,7 +128,34 @@ pub trait ContainerBuilder: Default + 'static {
 /// Specifically, the sum of lengths of all extracted and finished containers must equal the
 /// number of times that `push_into` is called on the container builder.
 /// If you have any questions about this trait you are best off not implementing it.
-pub trait LengthPreservingContainerBuilder : ContainerBuilder { }
+pub trait CountPreservingContainerBuilder: ContainerBuilder { }
```

Comment:
Names are a bit weird here, because "count" is a method on `WithProgress` and not obviously a top-level concept. I wonder if it makes sense to require all container builders to be count preserving? As in, you now need to implement both `count()` and `len()`, and there's no room for containers whose `count()` method is incorrect?

Reply:
Wasn't the idea of the trait to indicate that it's compatible with `partition`, i.e., after partitioning, the sum length of the parts equals the length of the input? From this perspective, it seems worth keeping around.

I restored the old name.

Reply:
It might still be useful for a container builder that compacts in-place, like we have in Differential. The consolidating container builder certainly can only be used before we fix counts for progress tracking. It can't be used for an exchange because data might consolidate differently.

I see two ways out: Either we have a container that stores the count independently of the contents, and sets the count to the number of `push_into` calls _outside_ progress tracking, or rework the exchange logic to make it clear that container builders inside have less flexibility than the ones before we take the count.

For this reason, I think it makes sense to leave this trait in Timely for the time being.